File: exercises/RB101_small_problems/advanced/madlibs.rb
=begin

Start time: 

# --------------------------- Problem
  Restate the problem:

  Make a madlibs program that reads in some text from a text file that you have created, and then plugs in a selection of randomized nouns, verbs, adjectives, and adverbs into that text and prints it. You can build your lists of nouns, verbs, adjectives, and adverbs directly into your program, but the text data should come from a file or other external source. Your program should read this text, and for each line, it should place random words of the appropriate types into the text, and print the result.

  Input: 
  Output:

# --------------------------- Test Cases + Rules
  Explicit:

  Implicit:

# --------------------------- Data Structure

# --------------------------- Scratch Book

# --------------------------- Algorithm

=end

ADJECTIVES = %w(quick lazy sleepy ugly).freeze
NOUNS = %w(fox dog head leg).freeze
VERBS = %w(jumps lifts bites licks).freeze
ADVERB = %w(easily lazily noisily excitedly).freeze

def word_substitute(word)
  case
  when word.include?('<adjective>') then word.gsub('<adjective>', ADJECTIVES.sample)
  when word.include?('<noun>') then word.gsub('<noun>', NOUNS.sample)
  when word.include?('<verb>') then word.gsub('<verb>', VERBS.sample)
  when word.include?('<adverb>') then word.gsub('<adverb>', ADVERB.sample)
  else
    word
  end
end

def madlibs(text)
  puts text.split.map { |word| word_substitute(word) }.join(' ')
end

text = "The <adjective> brown <noun> <adverb>
<verb> the <adjective> yellow
<noun>, who <adverb> <verb> his
<noun> and looks around."

madlibs(text)

File: exercises/JS210_Small_Problems/Advanced/madlibs.js
/*

Input: A template (most likely an object)
Output: A string of random nouns, adjectives, adverbs, and verbs

Example:
  Template 1:
    => "The NOUN VERB ADVERB the ADJECTIVE yellow NOUN, who ADVERB VERB his NOUN and looks around"
  
  Template 2:
    => "The NOUN VERB the NOUN's NOUN"

Rules:
  - The template:
    - Must contain a list of nouns, verbs, adverbs, and adjectives
    - Must contain a sentence structure
  - Return empty string if arg is omitted
  - Everytime the function is called, the sentence must contain random words

Data Structure (template):

  template = {
    structure = "The NOUN VERB the NOUN's NOUN",
    nouns: [],
    verbs: [],
    adverbs: [],
    adjectives: [],
  }

Algorithm:
1) Make parameter default to empty string if no arg is provided

2) Initialize variables nouns, verbs...

3) Assign each variable to a random value from the arrays
  - define a random number generator that returns a number from 0 to an arrays length
  - assign the returned value to the variable

4) Return the string structure from the object

*/
function getRandomWord(arr) {
  return arr[Math.floor(Math.random() * arr.length)]
}

function madlibs(template = '') {
  const nouns = 'fox dog head leg tail cat'.split(' ');
  const verbs = 'jumps lifts bites licks pats'.split(' ');
  const adverbs = 'easily lazily noisily excitedly'.split(' ');
  const adjectives = 'quick lazy sleepy noisy hungry'.split(' ');

  let sentence = template.replace(/NOUN/g, () => getRandomWord(nouns))
                         .replace(/ADVERB/g, () => getRandomWord(adverbs))
                         .replace(/VERB/g, () => getRandomWord(verbs))
                         .replace(/ADJECTIVE/g, () => getRandomWord(adjectives));

  return sentence;
}

const template1 = 'The ADJECTIVE brown NOUN ADVERB VERB the ADJECTIVE yellow NOUN, who ADVERB VERB his NOUN and looks around';

const template2 = "The NOUN VERB the NOUN's NOUN";

// These examples use the following list of replacement texts:
// adjectives: quick lazy sleepy noisy hungry
// nouns: fox dog head leg tail cat
// verbs: jumps lifts bites licks pats
// adverbs: easily lazily noisily excitedly
// ------

console.log(madlibs(template1));
// The "sleepy" brown "cat" "noisily"
// "licks" the "sleepy" yellow
// "dog", who "lazily" "licks" his
// "tail" and looks around.

console.log(madlibs(template1));
// The "hungry" brown "cat" "lazily"
// "licks" the "noisy" yellow
// "dog", who "lazily" "licks" his
// "leg" and looks around.

console.log(madlibs(template2));      // The "fox" "bites" the "dog"'s "tail".

console.log(madlibs(template2));      // The "cat" "pats" the "cat"'s "head".